File: mark/shopnsaveApify/shopnsave.js
const Apify = require('apify');
const { parseString } = require('xml2js');
const rp = require('request-promise-native');

// ENV URL: https://www.shopnsavefood.com/DesktopModules/StoreLocator/API/StoreWebAPI.asmx/GetAllStores
const storeurl = 'https://www.shopnsavefood.com/DesktopModules/StoreLocator/API/StoreWebAPI.asmx/GetAllStores';
const noDataLabel = 'NO-DATA';

// Converts XML to JSON
const xml2json = xmlString => new Promise((fullfill, reject) => {
  parseString(xmlString, { explicitArray: false }, (err, result) => {
    if (err) {
      reject(err);
    } else {
      fullfill(result);
    }
  });
});

// Makes the a phone number 10 digits with no punctutations
const formatPhoneNumber = string => string.replace(/\D/g, '');

const validateHours = (string1, string2) => {
  if (string1 === 'NO-DATA' || string1.length === 0) {
    return 'NO-DATA';
  }
  if (string1 !== 'NO-DATA' || string1.length !== 0) {
    if (string2 === 'NO-DATA' || string2.length === 0) {
      return string1;
    }
  }
  return `${string1}, ${string2}`;
};

// Simply receives data from the scrape, then formats it.
const parsedShopData = ({
  // If any data points are undefined / null, return 'NO-DATA'
  locatorDomain = 'shopnsavefood.com__api', Name = noDataLabel, Address1 = noDataLabel, City = noDataLabel,
  State = noDataLabel, Zip = noDataLabel, countryCode = 'US', StoreID = noDataLabel,
  Phone: Phone = noDataLabel, IsGasStation = noDataLabel,
  naics = noDataLabel, Latitude = noDataLabel,
  Longitude = noDataLabel, Hours = noDataLabel, Hours2 = noDataLabel,
}) => ({
    // Then set the label similar to the template and make adjustments (gas station / store / hours)
    locator_domain: locatorDomain,
    location_name: Name,
    street_address: Address1,
    city: City,
    state: State,
    zip: Zip,
    country_code: countryCode,
    store_number: StoreID,
    // Although redundant is required due to format Phone Number
    ...((Phone === 'NO-DATA') && { phone: 'NO-DATA' }),
    ...((Phone !== 'NO-DATA') && { phone: formatPhoneNumber(Phone) }),
    ...((IsGasStation === 'true') && { location_type: 'Gas Station' }),
    ...((IsGasStation === 'false') && { location_type: 'Store' }),
    naics_code: naics,
    latitude: Latitude,
    longitude: Longitude,
    hours_of_operation: validateHours(Hours, Hours2),
  });

Apify.main(async () => {
  const requestList = new Apify.RequestList({
    sources: [
      { url: storeurl },
    ],
  });
  await requestList.initialize();

  const crawler = new Apify.BasicCrawler({
    requestList,
    handleRequestFunction: async ({ request }) => {
      const html = await rp(request.url);
      const json = await xml2json(html);

      // The data is nested so define data to this new object
      const data = json.ArrayOfStore.Store;

      /* eslint-disable no-restricted-syntax */
      for await (const obj of data) {
        await Apify.pushData(parsedShopData(obj));
      }
    },
  });

  await crawler.run();
});

File: archive/darin/shopnsave.py
import urllib2
import time
import os
import re
import csv

path = '.'  ### this is a relative path for where the output file will be put

header = 'locator_domain,location_name,street_address,city,state,zip,country_code,store_number,phone,location_type,naics_code,latitude,longitude,hours_of_operation\n'
with open("%s/ShopNSave.csv" % path, "wb") as out_file:
            out_file.write(header)

domain = 'https://www.shopnsavefood.com'

url = 'https://www.shopnsavefood.com/DesktopModules/StoreLocator/API/StoreWebAPI.asmx/GetAllStores'
page_text = urllib2.urlopen(url)
for line in page_text:
    if '<StoreID>' in line:
        hid = line.split('>')[1].split('<')[0].replace('"',"'")
        name = ''
        add = ''
        city = ''
        zc = ''
        state = ''
        phone = ''
        lat = ''
        lng = ''
        hrs = ''
        brand = ''
    if '<Name>' in line:
        name = line.split('>')[1].split('<')[0].replace('"',"'")
    if '<Address1>' in line:
        add = line.split('>')[1].split('<')[0].replace('"',"'")
    if '<City>' in line:
        city = line.split('>')[1].split('<')[0].replace('"',"'")
    if '<State>' in line:
        state = line.split('>')[1].split('<')[0].replace('"',"'")
    if '<Zip>' in line:
        zc = line.split('>')[1].split('<')[0].replace('"',"'")
    country = 'US'
    if '<Phone>' in line:
        phone = line.split('>')[1].split('<')[0].replace('"',"'")
    if '<Latitude>' in line:
        lat = line.split('>')[1].split('<')[0].replace('"',"'")
    if '<Longitude>' in line:
        lng = line.split('>')[1].split('<')[0].replace('"',"'")
    if '<Hours>' in line:
        hrs = line.split('>')[1].split('<')[0].replace('"',"'")
    if '<Hours2>' in line:
        hrs = hrs + ', ' + line.split('>')[1].split('<')[0].replace('"',"'")
    if '<IsGasStation>true' in line:
        brand = 'Fuel'
    if '<IsGasStation>false' in line:
        brand = 'Store'
    if '</Store>' in line:
        info = '"' + domain + '","' + name + '","' + add + '","' + city + '","' + state + '","' + zc + '","' + country
        info = info + '","' + hid + '","' + phone + '","' + brand + '","","' + lat + '","' + lng + '","' + hrs + '"'
        info = info.replace('\n','').replace('\r','').replace('&amp;','&')
        with open("%s/ShopNSave.csv" % path, "ab") as out_file:
            out_file.write(info + '\n')
